File: src/foam/core/CompoundContextAgency.java
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.core;

import java.util.ArrayList;

/**
 * A ContextAgent which is itself an Agency.
 * When ContextAgents are submit()'ed to the service it queues them
 * but doesn't execute them until it is itself executed.
 **/
public class CompoundContextAgency
  implements ContextAgent, Agency
{
  public ArrayList<Runnable> agents_ = new ArrayList();

  public void execute(X x) {
    CompoundException e = new CompoundException();
    for ( Runnable agent : agents_ ) {
      try {
        agent.run();
      } catch (Throwable t) {
        e.exceptions.add(t);
      }
    }
    if ( e.exceptions.size() != 0 ) {
      throw e;
    }
  }

  public void submit(X x, ContextAgent agent, String description) {
    agents_.add(new ContextAgentRunnable(x, agent, description));
  }
  public void submit(X x, ContextAgent agent) {
    submit(x, agent, "");
  }

  public String toString() {
    StringBuilder sb = new StringBuilder();
    boolean first = true;
    for ( Runnable agent : agents_ ) {
      if ( first ) {
        first = false;
      } else {
        sb.append("\n");
      }
      sb.append(agent.toString());
    }

    return sb.toString();
  }

  public String[] describe() {
    String[] desc = new String[agents_.size()];

    for ( int i = 0 ; i < agents_.size() ; i++ ) {
      desc[i] = agents_.get(i).toString();
    }

    return desc;
  }
}

File: src/foam/core/CompoundContextAgency.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.core',
  name: 'CompoundContextAgency',
  javaImplements: [
    'foam.core.Agency',
    'foam.core.ContextAgent'
  ],

  javaImports: [
    'foam.nanos.logger.Logger',
    'java.util.ArrayList'
  ],

  properties: [
    {
      name: 'agents',
      class: 'List',
      javaType: 'ArrayList<Runnable>',
      javaFactory: `return new ArrayList();`
    }
  ],

  methods: [
    {
      name: 'execute',
      args: [
        { name: 'x', type: 'X' }
      ],
      javaCode: `CompoundException e = new CompoundException();
Logger logger = (Logger) x.get("logger");
for ( Runnable agent : getAgents() ) {
  try {
    agent.run();
  } catch (Throwable t) {
    logger.error(t.getMessage(), t);
    e.add(t);
  }
}
e.maybeThrow();`
    },
    {
      name: 'submit',
      args: [
        { name: 'x', type: 'X' },
        { name: 'agent', type: 'ContextAgent' },
        { name: 'description', type: 'String' }
      ],
      javaCode: `getAgents().add(new ContextAgentRunnable(x, agent, description));`
    },
    {
      name: 'toString',
      type: 'String',
      javaCode: `StringBuilder sb = new StringBuilder();
boolean first = true;
for ( Runnable agent : getAgents() ) {
  if ( first ) {
    first = false;
  } else {
    sb.append(System.lineSeparator());
  }
  sb.append(agent.toString());
}
return sb.toString();`
    },
    {
      name: 'describeAgents',
      type: 'String[]',
      javaCode: `String[] desc = new String[getAgents().size()];
for ( int i = 0 ; i < getAgents().size() ; i++ ) {
  desc[i] = getAgents().get(i).toString();
}
return desc;`
    }
  ],
  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(`public void submit(foam.core.X x, foam.core.ContextAgent agent) {
          submit(x, agent, "");
        }`);
      }
    }
  ]
});

